### PR TITLE
recycle buffers in asd_client

### PR DESCRIPTION
**Pull request by toolslive**

### CODE REVIEW

File: ocaml/src/tools/llio2.ml [hunk 1]
```diff
@@ -21,22 +21,26 @@ open Lwt_bytes2
...
+               mutable pos : int;
+               too_far : int;
+             }
 
     type 'a deserializer = t -> 'a
 
-    let make_buffer buf pos = { buf; pos; }
+    let make_buffer buf pos too_far = { buf; pos; too_far }
 
     let advance_pos buf delta = buf.pos <- buf.pos + delta
```
Comment (domsj) [inline ocaml/src/tools/llio2.ml]: add an assert here that `buf.pos <= buf.max_pos`
(max_pos is more descriptive for me than too_far)

Comment (toolslive) [inline ocaml/src/tools/llio2.ml]: ok

File: ocaml/src/tools/llio2.ml [hunk 2]
```diff
@@ -21,22 +21,28 @@ open Lwt_bytes2
...
+             }
 
     type 'a deserializer = t -> 'a
 
-    let make_buffer buf pos = { buf; pos; }
+    let make_buffer buf pos max_pos =
+      assert (pos <= max_pos);
+      { buf; pos; max_pos }
 
     let advance_pos buf delta = buf.pos <- buf.pos + delta
```
Comment (domsj) [inline ocaml/src/tools/llio2.ml]: you still want to add an assert here too

**Comment by domsj** [comment]
it probably makes sense to release (unsafe_destroy) the buffer when the client is disposed, otherwise we'll introduce a space leak

**Comment by domsj** [comment]
ok

**Pull request merged**